spinalcordtoolbox/labels.py

```diff
@@ -224,52 +227,46 @@ def label_regions_from_reference(img: Image, ref: Image, centerline: bool = Fals
     :param centerline: boolean, if True the centerline is returned instead of the segmentation (default to false)
     :returns: segmentation image with vertebral levels labelized
     """
-
-    og_img_orientation = img.orientation
-    if og_img_orientation != "RPI":
-        img.change_orientation("RPI")
-
-    og_ref_orientation = ref.orientation
-    if og_ref_orientation != "RPI":
-        ref.change_orientation("RPI")
-
-    out = zeros_like(img)
-
-    # Extract centerline from segmentation
-    _, arr_ctl, _, _ = get_centerline(img)
-    centerline_arr = arr_ctl.T
-
+    # The centerline of the segmentation, with one point for each S-I slice.
+    centerline_pix = get_centerline(img, space='pix')[1].T.astype(int)  # shape (N, 3), int
+    centerline_phys = img.transfo_pix2phys(centerline_pix, mode='absolute')  # shape (N, 3), float
+
+    # Prepare to compute, for several points, the centerline index which is closest in physical space.
+    centerline_tree = KDTree(centerline_phys)
+
+    # The label coordinates.
+    labels_pix = np.argwhere(ref.data)  # shape (L, 3), int
+    labels_phys = ref.transfo_pix2phys(labels_pix, mode='absolute')  # shape (L, 3), float
+    # The corresponding label values.
+    labels_values = ref.data[tuple(labels_pix.T)]  # shape (L,), arbitrary dtype
+    # The closest centerline index for each label voxel.
+    _, labels_indices = centerline_tree.query(labels_phys)  # shape (L,), int
+
+    # Sort everything by increasing centerline index
+    sorter = np.argsort(labels_indices)
+    labels_pix = labels_pix[sorter]
+    labels_phys = labels_phys[sorter]
+    labels_values = labels_values[sorter]
+    labels_indices = labels_indices[sorter]
+    del sorter
+
+    # Add a label for above the top-most label
+    labels_values = list(labels_values)
+    labels_values.append(labels_values[-1] - 1)
```

Comment:
Side note: Now that I look at it... if the top-most label is `1`, then the label for the region above `1` will be `0`. 

On the one hand, I think this is somewhat reasonable (since capping it at `1` would cause label-based aggregation to be wrong, and using a non-sequential dummy value like `256` might cause some sort of unexpected error by breaking the assumption that labels are monotonically decreasing).  

But at the same time, outputting _no_ metrics for this case also doesn't feel great? Hrmmm...

I guess this preserves the old behavior, so we can always just open an issue to address it at a later point.